pkg/graveler/settings/manager.go

```diff
@@ -110,54 +125,51 @@ func (m *Manager) Get(ctx context.Context, repositoryID graveler.RepositoryID, k
-		StorageNamespace: string(repo.StorageNamespace),
-		Identifier:       fmt.Sprintf(settingsRelativeKey, m.committedBlockStoragePrefix, key),
-		IdentifierType:   block.IdentifierTypeRelative,
-	}
-	reader, err := m.blockAdapter.Get(ctx, objectPointer, -1)
-	if errors.Is(err, adapter.ErrDataNotFound) {
-		return nil, graveler.ErrNotFound
-	}
-	if err != nil {
-		return nil, err
-	}
-	defer func() {
-		_ = reader.Close()
-	}()
-	return io.ReadAll(reader)
-}
-
-// UpdateWithLock atomically gets a setting, performs the update function, and persists the setting to the store.
+// Update atomically gets a setting, performs the update function, and persists the setting to the store.
 // The settingTemplate parameter is used to determine the type passed to the update function.
-func (m *Manager) UpdateWithLock(ctx context.Context, repositoryID graveler.RepositoryID, key string, settingTemplate proto.Message, update func(proto.Message) error) error {
-	repo, err := m.refManager.GetRepository(ctx, repositoryID)
-	if err != nil {
-		return err
-	}
-	_, err = m.branchLock.MetadataUpdater(ctx, repositoryID, repo.DefaultBranchID, func() (interface{}, error) {
-		setting, err := m.GetLatest(ctx, repositoryID, key, settingTemplate)
+func (m *KVManager) Update(ctx context.Context, repositoryID graveler.RepositoryID, key string, settingTemplate proto.Message, update updateFunc) error {
+	const (
+		maxInterval = 2
+		maxElapsed  = 5
+	)
+	bo := backoff.NewExponentialBackOff()
+	bo.MaxInterval = maxInterval * time.Second
+	bo.MaxElapsedTime = maxElapsed * time.Second
+
+	err := backoff.Retry(func() error {
+		repo, err := m.refManager.GetRepository(ctx, repositoryID)
+		if err != nil {
+			return backoff.Permanent(err)
+		}
+
+		data := settingTemplate.ProtoReflect().Interface()
+		pred, err := m.getWithPredicate(ctx, &graveler.RepositoryRecord{RepositoryID: repositoryID, Repository: repo}, key, data)
 		if errors.Is(err, graveler.ErrNotFound) {
-			setting = proto.Clone(settingTemplate)
+			data = proto.Clone(settingTemplate)
 		} else if err != nil {
-			return nil, err
+			return backoff.Permanent(err)
 		}
-		logSetting(logging.FromContext(ctx), repositoryID, key, setting, "got repository-level setting")
-		err = update(setting)
+
+		logSetting(logging.FromContext(ctx), repositoryID, key, data, "update repository-level setting")
+		err = update(data)
 		if err != nil {
-			return nil, err
+			return backoff.Permanent(err)
 		}
-		return nil, m.Save(ctx, repositoryID, key, setting)
-	})
+		err = m.store.SetMsgIf(ctx, graveler.RepoPartition(&graveler.RepositoryRecord{RepositoryID: repositoryID, Repository: repo}), []byte(graveler.SettingsPath(key)), data, pred)
+		if errors.Is(err, kv.ErrPredicateFailed) {
+			err = graveler.ErrPreconditionFailed
+		} else if err != nil {
+			return backoff.Permanent(err)
+		}
+		return err
+	}, bo)
+	if bo.NextBackOff() == bo.Stop {
+		return fmt.Errorf("update settings: %w", graveler.ErrTooManyTries)
```

Comment:
Do we log the errors that occur? If we maxed our retries with outputing any indication for the reason then it's hard for the user/developer to understand what went wrong..